File: tools/vault/root_anchor.py
"""A stable vault recipient reached through the personal root.

The anchor is deliberately separate from both kinds of key the personal root
already has:

* the Ed25519 root remains a signing identity;
* a purpose-derived X25519 recipient, derived from the root seed, opens one
  random 32-byte anchor seed;
* that anchor seed derives a policy-recipient X25519 keypair to which a class
  can wrap its class key. Organization personas use the same recipient seam
  under a different purpose-bound kind.

Consequently root-factor churn changes only the armor.  It never changes this
record or any policy class.  A root rotation re-seals this one anchor seed to
the new root-derived recipient; individual Settings remain untouched.
"""

from __future__ import annotations

import hashlib
import secrets
from dataclasses import dataclass

from tools.network.idkit.canonical import canonical_json
from tools.network.idkit.keys import KeyPair, verify_signature
from tools.network.idkit.sealing import (
    derive_encapsulation_keypair,
    open as seal_open,
    seal,
)

from .errors import FactorError, VaultError
from .recipients import (
    PERSONAL_ROOT_RECIPIENT,
    PublishedRecipient,
    recipient_public_from_seed,
)


ROOT_ANCHOR_WRAP_PURPOSE = "autonomy/vault-root-anchor-wrap/v1"
ROOT_ANCHOR_ENROLL_DOMAIN = b"autonomy/vault-root-anchor-enroll/v1\n"
ROOT_ANCHOR_VERSION = 1
_SEED_LEN = 32
_HEX_32 = 64
# RFC 9180 suite byte + X25519 encapsulated key + 32-byte plaintext + tag.
_SEALED_SEED_HEX_LEN = (1 + 32 + _SEED_LEN + 16) * 2


def _require_hex32(value: object, label: str) -> str:
    if not isinstance(value, str) or len(value) != _HEX_32:
        raise VaultError(f"root anchor {label} must be 64 lowercase hex characters")
    if any(ch not in "0123456789abcdef" for ch in value):
        raise VaultError(f"root anchor {label} must be 64 lowercase hex characters")
    return value


def wrap_purpose(anchor_id: str, root_pub: str) -> str:
    """Bind a sealed seed to one anchor and one personal-root identity."""
    digest = hashlib.sha256(canonical_json({
        "anchor_id": anchor_id,
        "root_pub": root_pub,
        "v": ROOT_ANCHOR_VERSION,
    })).hexdigest()
    return f"{ROOT_ANCHOR_WRAP_PURPOSE}|{digest}"


@dataclass(frozen=True)
class RootAnchorRecord:
    """The public, root-attested envelope stored in the personal database."""

    anchor_id: str
    display_name: str
    root_pub: str
    public_key: str
    sealed_seed: str
    created_at: str
    signature: str
    v: int = ROOT_ANCHOR_VERSION

    def unsigned_dict(self) -> dict:
        return {
            "v": self.v,
            "anchor_id": self.anchor_id,
            "display_name": self.display_name,
            "root_pub": self.root_pub,
            "public_key": self.public_key,
            "sealed_seed": self.sealed_seed,
            "created_at": self.created_at,
        }

    def to_dict(self) -> dict:
        return {**self.unsigned_dict(), "signature": self.signature}

    @classmethod
    def from_dict(cls, value: dict) -> "RootAnchorRecord":
        if not isinstance(value, dict) or set(value) != {
            "v", "anchor_id", "display_name", "root_pub", "public_key",
            "sealed_seed", "created_at", "signature",
        }:
            raise VaultError("root anchor record has unknown or missing fields")
        if value.get("v") != ROOT_ANCHOR_VERSION:
            raise VaultError(f"unsupported root anchor version {value.get('v')!r}")
        for field in ("anchor_id", "display_name", "created_at"):
            member = value.get(field)
            if not isinstance(member, str) or not member or len(member) > 256:
                raise VaultError(f"root anchor {field} must be a short non-empty string")
        _require_hex32(value.get("root_pub"), "root_pub")
        _require_hex32(value.get("public_key"), "public_key")
        sealed_seed = value.get("sealed_seed")
        if (
            not isinstance(sealed_seed, str)
            or len(sealed_seed) != _SEALED_SEED_HEX_LEN
            or any(ch not in "0123456789abcdef" for ch in sealed_seed)
        ):
            raise VaultError("root anchor sealed_seed is not a canonical HPKE record")
        signature = value.get("signature")
        if (
            not isinstance(signature, str)
            or len(signature) != 128
            or any(ch not in "0123456789abcdef" for ch in signature)
        ):
            raise VaultError("root anchor signature must be 128 lowercase hex characters")
        record = cls(**value)
        try:
            verify_signature(
                record.root_pub,
                record.signature,
                ROOT_ANCHOR_ENROLL_DOMAIN + canonical_json(record.unsigned_dict()),
            )
        except Exception as exc:
            raise VaultError("root anchor is not signed by its personal root") from exc
        return record

    def published_recipient(self) -> PublishedRecipient:
        return PublishedRecipient(
            self.anchor_id, PERSONAL_ROOT_RECIPIENT, self.public_key,
        )


def create_root_anchor(
    root: KeyPair,
    *,
    anchor_id: str,
    display_name: str,
    created_at: str,
    anchor_seed: bytes | None = None,
) -> tuple[RootAnchorRecord, bytes]:
    """Create and root-sign an anchor envelope (Python harness/reference)."""
    seed = bytes(anchor_seed if anchor_seed is not None else secrets.token_bytes(_SEED_LEN))
    if len(seed) != _SEED_LEN:
        raise FactorError("root anchor seed must be exactly 32 bytes")
    purpose = wrap_purpose(anchor_id, root.public_hex)
    _root_private, root_recipient = derive_encapsulation_keypair(
        bytes.fromhex(root.private_hex), purpose,
    )
    unsigned = {
        "v": ROOT_ANCHOR_VERSION,
        "anchor_id": anchor_id,
        "display_name": display_name,
        "root_pub": root.public_hex,
        "public_key": recipient_public_from_seed(seed, PERSONAL_ROOT_RECIPIENT),
        "sealed_seed": seal(
            seed, root_recipient, purpose,
        ).hex(),
        "created_at": created_at,
    }
    signature = root.sign_hex(ROOT_ANCHOR_ENROLL_DOMAIN + canonical_json(unsigned))
    return RootAnchorRecord.from_dict({**unsigned, "signature": signature}), seed


def open_root_anchor(record: RootAnchorRecord, root_seed: bytes) -> bytes:
    """Open *record* after a ceremony produced the matching root seed."""
    purpose = wrap_purpose(record.anchor_id, record.root_pub)
    root_private, _root_public = derive_encapsulation_keypair(root_seed, purpose)
    try:
        seed = seal_open(
            bytes.fromhex(record.sealed_seed),
            root_private,
            purpose,
        )
    except Exception as exc:
        raise VaultError("the opened personal root does not open this vault anchor") from exc
    if (
        len(seed) != _SEED_LEN
        or recipient_public_from_seed(seed, PERSONAL_ROOT_RECIPIENT)
        != record.public_key
    ):
        raise VaultError("the root anchor seed does not match its published recipient")
    return seed

File: tools/dashboard/static/js/ceremony/root-anchor.js
/* Personal-root vault anchor, byte-compatible with tools/vault/root_anchor.py.
 *
 * Passwords and passkeys stop at opening the personal root armor. This module
 * turns that freshly opened root into the stable recipient seed used by a
 * root-reachable policy class. Neither the root seed nor the anchor seed is
 * persisted or sent except for the anchor seed's one approval-bound opener.
 */
import {
  bytesToHex,
  canonicalJson,
  deriveEncapsulationKeypair,
  hexToBytes,
  openSealedArmor,
  sealToEncapsulationKey,
} from './primitives.js';

export const PERSONAL_ROOT_RECIPIENT = 'personal-root-anchor';
export const POLICY_RECIPIENT_PURPOSE = 'autonomy/vault-policy-recipient/v1';
export const ROOT_ANCHOR_WRAP_PURPOSE = 'autonomy/vault-root-anchor-wrap/v1';
export const ROOT_ANCHOR_ENROLL_DOMAIN = 'autonomy/vault-root-anchor-enroll/v1\n';

const encoder = new TextEncoder();

function concat(left, right) {
  const out = new Uint8Array(left.length + right.length);
  out.set(left, 0); out.set(right, left.length);
  return out;
}

async function sha256Hex(bytes) {
  return bytesToHex(new Uint8Array(await crypto.subtle.digest('SHA-256', bytes)));
}

export async function rootAnchorWrapPurpose(anchorId, rootPub) {
  const digest = await sha256Hex(encoder.encode(canonicalJson({
    anchor_id: anchorId,
    root_pub: rootPub,
    v: 1,
  })));
  return `${ROOT_ANCHOR_WRAP_PURPOSE}|${digest}`;
}

function recipientPurpose(kind) {
  return `${POLICY_RECIPIENT_PURPOSE}|${kind}`;
}

export async function createRootAnchorEnvelope(
  openedRoot,
  {
    anchorId = 'personal-root-default',
    displayName = 'Personal root vault access',
    createdAt = new Date().toISOString(),
    anchorSeed = null,
  } = {},
) {
  if (!openedRoot?.seed || !openedRoot?.signingKey || !openedRoot?.rootPub) {
    throw new Error('a freshly opened personal root is required');
  }
  const seed = anchorSeed
    ? new Uint8Array(anchorSeed)
    : crypto.getRandomValues(new Uint8Array(32));
  if (seed.length !== 32) {
    seed.fill(0);
    throw new Error('root anchor seed must be exactly 32 bytes');
  }
  try {
    const purpose = await rootAnchorWrapPurpose(anchorId, openedRoot.rootPub);
    const rootRecipient = await deriveEncapsulationKeypair(openedRoot.seed, purpose);
    const anchorRecipient = await deriveEncapsulationKeypair(
      seed, recipientPurpose(PERSONAL_ROOT_RECIPIENT),
    );
    const sealed = await sealToEncapsulationKey(
      seed, rootRecipient.publicKeyHex, purpose,
    );
    const unsigned = {
      v: 1,
      anchor_id: anchorId,
      display_name: displayName,
      root_pub: openedRoot.rootPub,
      public_key: anchorRecipient.publicKeyHex,
      sealed_seed: bytesToHex(sealed),
      created_at: createdAt,
    };
    const signature = bytesToHex(new Uint8Array(await crypto.subtle.sign(
      'Ed25519',
      openedRoot.signingKey,
      concat(
        encoder.encode(ROOT_ANCHOR_ENROLL_DOMAIN),
        encoder.encode(canonicalJson(unsigned)),
      ),
    )));
    return { ...unsigned, signature };
  } finally {
    seed.fill(0);
  }
}

export async function openRootAnchorEnvelope(anchor, rootSeed) {
  const fields = [
    'anchor_id', 'created_at', 'display_name', 'public_key', 'root_pub',
    'sealed_seed', 'signature', 'v',
  ];
  if (!anchor || typeof anchor !== 'object'
      || Object.keys(anchor).sort().join('|') !== fields.sort().join('|')
      || anchor.v !== 1 || typeof anchor.anchor_id !== 'string'
      || typeof anchor.display_name !== 'string' || typeof anchor.created_at !== 'string'
      || typeof anchor.root_pub !== 'string' || !/^[0-9a-f]{64}$/.test(anchor.root_pub)
      || typeof anchor.public_key !== 'string' || !/^[0-9a-f]{64}$/.test(anchor.public_key)
      || typeof anchor.sealed_seed !== 'string' || !/^[0-9a-f]+$/.test(anchor.sealed_seed)
      || typeof anchor.signature !== 'string' || !/^[0-9a-f]{128}$/.test(anchor.signature)) {
    throw new Error('root anchor envelope is malformed');
  }
  const unsigned = {
    v: anchor.v,
    anchor_id: anchor.anchor_id,
    display_name: anchor.display_name,
    root_pub: anchor.root_pub,
    public_key: anchor.public_key,
    sealed_seed: anchor.sealed_seed,
    created_at: anchor.created_at,
  };
  const verificationKey = await crypto.subtle.importKey(
    'raw', hexToBytes(anchor.root_pub), { name: 'Ed25519' }, false, ['verify'],
  );
  const signatureValid = await crypto.subtle.verify(
    'Ed25519',
    verificationKey,
    hexToBytes(anchor.signature),
    concat(
      encoder.encode(ROOT_ANCHOR_ENROLL_DOMAIN),
      encoder.encode(canonicalJson(unsigned)),
    ),
  );
  if (!signatureValid) {
    throw new Error('root anchor is not signed by this personal root');
  }
  const purpose = await rootAnchorWrapPurpose(anchor.anchor_id, anchor.root_pub);
  const seed = await openSealedArmor({
    sealed_root_key: anchor.sealed_seed,
    seal_purpose: purpose,
  }, rootSeed);
  try {
    if (seed.length !== 32) throw new Error('root anchor opened malformed material');
    const recipient = await deriveEncapsulationKeypair(
      seed, recipientPurpose(PERSONAL_ROOT_RECIPIENT),
    );
    if (recipient.publicKeyHex !== anchor.public_key) {
      throw new Error('root anchor seed does not match its published recipient');
    }
    return new Uint8Array(seed);
  } finally {
    seed.fill(0);
  }
}
